routing/DELEGATED_NAMING_HTTP.md

```diff
@@ -0,0 +1,176 @@
+  unless stated otherwise.
+- `Payload` is the content of the IPNS record as specified
+  by [IPNS Record](../ipns/IPNS.md#ipns-record) specification:
+- `Value` is the string representation of the IPNS path,
+  e.g. `ipns/{ipns-key}`, `/ipns/example.com`, `/ipfs/baf...`, etc.
+- `Sequence` represents the current version of the record starting from `0`.
+- `Validity` captures the mechanism by which the record is validated. Each validity type reserves a
+  field key under this object.
+  - `EOL` donates that the validity type is EOL, containing:
+    - `Timestamp` represents the time in the future at which the record expires with nanoseconds
+       precision represented as an ASCII string that follows notation
+       from [RFC3339](https://www.ietf.org/rfc/rfc3339.txt).
+    - `AdvisoryTTL` represents an optional field that hints at how long the record should be
+       cached.
+- `PublicKey` represents the optional public key used to sign the record. This field is only
+  required if it cannot be extracted from the IPNS name, e.g. in the case of legacy RSA keys.
+- `ExtendedData` represents the extensible data as arbitrary JSON object.
+
+## Versioning
+
+The path predix `/v1` donates the version number of the HTTP API. Backwards-incompatible change must
+increment the version number.
+
+## API
+
+### `GET /naming/v1/records/{ipns-name}`
+
+**Path Parameters**
+
+- `ipns-name` the IPNS name to resolve.
+
+**Response Status Codes**
+
+- `200` (OK): indicates that the response body containing the IPNS record that corresponds to the
+  IPNS name.
+- `404` (Not Found): indicates that no matching records are found.
+- `400` (Bad Request): indicates that the given IPNS name is not valid.
+- `429` (Too Many Requests): indicates that the caller is issuing requests too many request and may
+  retry after the time specified
+  at [Retry-After](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After) response
+  header has elapsed.
+- `501` (Not Implemented): indicates that the server does not support resolution of IPNS records
+
+**Response Body**
+
+The response body contains the [IPNS record](#ipns-record).
+
+```json
+{
+  "Signature": "<signature>",
+  "Payload": {
+    "Value": "<value>",
+    "Sequence": 0,
+    "Validity": {
+      "EOL": {
+        "Timestamp": 0,
+        "AdvisoryTTL": 0
+      }
+    },
+    "PublicKey": "<optional-public-key>",
+    "ExtendedData": {}
+  }
+}
+```
+
+### `PUT /naming/v1/records/{ipns-name}`
+
+**Path Parameters**
+
+- `ipns-name` the IPNS name to publish. The name must match `Value` in request body.
+
+**Request Body**
+
+```json
+{
+  "Signature": "<signature>",
+  "Payload": {
+    "Value": "<value>",
+    "Sequence": 0,
+    "Validity": {
+      "EOL": {
+        "Timestamp": 0,
+        "AdvisoryTTL": 0
+      }
+    },
+    "PublicKey": "<optional-public-key>",
+    "ExtendedData": {}
+  }
+}
+```
```

Comment:
`v2` means something in the iteration of protobuf format; i am not clear why it should leak here. It's a detail that's IMHO is irrelevant from the perspective of the users whom would be using this API :) WDYT?

Reply:
What happens, when we decide to evolve IPNS in the future, and introduce signatureV3, and have both V2 and V3  for a while, to allow clients to do graceful upgrade?
Will you break existing users by suddenly swapping the value in `signature` field?
I don't think there is a good or bad answer here, just trade-offs. 

Would be good to add a paragraph to the IPIP saying "Signature field in JSON notation will always point at the latest version of signature from IPNS Record"


After thinking about it, I don't feel strongly about the details of JSON format, as long we also require endpoint to support binary `application/vnd.ipfs.ipns-record`.  Clients who care about signature verification will always use the latter anyway, because they would recreate most of it for signature verification anyway, so it is less expensive to work on it from start, than convert JSON  protobuf/cbor.